src/data/matrix/basic.lean

```diff
@@ -799,6 +811,114 @@ end
 
 end update
 
+section block_matrices
+
+/-- We can form a single large matrix by flattening smaller 'block' matrices of compatible
+dimensions. -/
+def from_blocks (A : matrix n l α) (B : matrix n m α) (C : matrix o l α) (D : matrix o m α) :
+  matrix (n ⊕ o) (l ⊕ m) α :=
+sum.elim (λ i, sum.elim (A i) (B i))
+         (λ i, sum.elim (C i) (D i))
+
+@[simp] lemma from_blocks_apply₁₁
+  (A : matrix n l α) (B : matrix n m α) (C : matrix o l α) (D : matrix o m α) (i : n) (j : l) :
+  from_blocks A B C D (sum.inl i) (sum.inl j) = A i j :=
+rfl
+
+@[simp] lemma from_blocks_apply₁₂
+  (A : matrix n l α) (B : matrix n m α) (C : matrix o l α) (D : matrix o m α) (i : n) (j : m) :
+  from_blocks A B C D (sum.inl i) (sum.inr j) = B i j :=
+rfl
+
+@[simp] lemma from_blocks_apply₂₁
+  (A : matrix n l α) (B : matrix n m α) (C : matrix o l α) (D : matrix o m α) (i : o) (j : l) :
+  from_blocks A B C D (sum.inr i) (sum.inl j) = C i j :=
+rfl
+
+@[simp] lemma from_blocks_apply₂₂
+  (A : matrix n l α) (B : matrix n m α) (C : matrix o l α) (D : matrix o m α) (i : o) (j : m) :
+  from_blocks A B C D (sum.inr i) (sum.inr j) = D i j :=
+rfl
+
+/-- Given a matrix whose row and column indexes are sum types, we can extract the correspnding
+"top left" submatrix. -/
+def to_blocks₁₁ (M : matrix (n ⊕ o) (l ⊕ m) α) : matrix n l α :=
+λ i j, M (sum.inl i) (sum.inl j)
+
+/-- Given a matrix whose row and column indexes are sum types, we can extract the correspnding
+"top right" submatrix. -/
+def to_blocks₁₂ (M : matrix (n ⊕ o) (l ⊕ m) α) : matrix n m α :=
+λ i j, M (sum.inl i) (sum.inr j)
+
+/-- Given a matrix whose row and column indexes are sum types, we can extract the correspnding
+"bottom left" submatrix. -/
+def to_blocks₂₁ (M : matrix (n ⊕ o) (l ⊕ m) α) : matrix o l α :=
+λ i j, M (sum.inr i) (sum.inl j)
+
+/-- Given a matrix whose row and column indexes are sum types, we can extract the correspnding
+"bottom right" submatrix. -/
+def to_blocks₂₂ (M : matrix (n ⊕ o) (l ⊕ m) α) : matrix o m α :=
+λ i j, M (sum.inr i) (sum.inr j)
+
+lemma from_blocks_to_blocks (M : matrix (n ⊕ o) (l ⊕ m) α) :
+  M = from_blocks M.to_blocks₁₁ M.to_blocks₁₂ M.to_blocks₂₁ M.to_blocks₂₂ :=
```

Comment:
Presumably we should also have a `to_blocks_from_blocks` (I guess four of them?)

Reply:
Agreed and done.